src/libbluechi/test/common/cfg/cfg_load_complete_configuration_test.c

```diff
@@ -0,0 +1,263 @@
+        char *key;
+        char *value;
+} cfg_test_expected_entry;
+
+typedef struct cfg_test_param {
+        char *test_case_name;
+        cfg_test_input input;
+
+        int expected_ret;
+        cfg_test_expected_entry expected_entries[MAX_EXPECTED_ENTRIES];
+} cfg_test_param;
+
+struct cfg_test_param test_data[] = {
+        /* basic single config file tests */
+        { "no config files",
+          { NULL, NULL, { NULL, NULL } },
+          0,
+          { { NULL, NULL }, { NULL, NULL }, { NULL, NULL } } },
+        { "only default config file",
+          { "NodeName = laptop\n", NULL, { NULL, NULL } },
+          0,
+          { { "NodeName", "laptop" }, { NULL, NULL }, { NULL, NULL } } },
+        { "only custom config file",
+          { NULL, "NodeName = laptop\n", { NULL, NULL } },
+          0,
+          { { "NodeName", "laptop" }, { NULL, NULL }, { NULL, NULL } } },
+        { "only file in confd",
+          { NULL, "NodeName = laptop\n", { "NodeName = laptop\n", NULL } },
+          0,
+          { { "NodeName", "laptop" }, { NULL, NULL }, { NULL, NULL } } },
+
+        /* testing of loading order and overwrite */
+        { "with custom config file",
+          { "NodeName = laptop\n", "NodeName = pi\n", { NULL, NULL } },
+          0,
+          { { "NodeName", "pi" }, { NULL, NULL }, { NULL, NULL } } },
+        { "with config file in confd",
+          { "NodeName = laptop\n", "NodeName = pi\n", { "NodeName = vm\n", NULL } },
+          0,
+          { { "NodeName", "vm" }, { NULL, NULL }, { NULL, NULL } } },
+        { "with multiple config files in confd",
+          { "NodeName = laptop\n", "NodeName = pi\n", { "NodeName = vm\n", "NodeName = container\n" } },
+          0,
+          { { "NodeName", "container" }, { NULL, NULL }, { NULL, NULL } } },
+
+        /* testing of loading order, overwrite and merges */
+        { "with multiple, duplicate keys",
+          { "NodeName = laptop\nLogLevel=INFO\n",
+            "NodeName = pi\n",
+            { "LogLevel=DEBUG\n", "LogTarget=stderr\n" } },
+          0,
+          { { "NodeName", "pi" }, { "LogLevel", "DEBUG" }, { "LogTarget", "stderr" } } },
+
+        /* testing invalid entries */
+        { "with invalid value in default config file",
+          { "NodeName 0 laptop\n", "NodeName = pi\n", { "LogLevel=DEBUG\n", "LogTarget=stderr\n" } },
+          -EINVAL,
+          { { NULL, NULL }, { NULL, NULL }, { NULL, NULL } } },
+        { "with invalid value in custom config file",
+          { "NodeName = laptop\n", "NodeName 0 pi\n", { "LogLevel=DEBUG\n", "LogTarget=stderr\n" } },
+          -EINVAL,
+          { { "NodeName", "laptop" }, { NULL, NULL }, { NULL, NULL } } },
+        { "with invalid value in one of the confd files",
+          { "NodeName = laptop\n", "NodeName = pi\n", { "LogLevel=DEBUG\n", "LogTarget0stderr\n" } },
+          -EINVAL,
+          { { "NodeName", "pi" }, { "LogLevel", "DEBUG" }, { NULL, NULL } } },
+};
```

Comment:
The table-driven approach is rather hard to understand here, but it still simplifies the actual test functions in my point of view. If not desired, I can create dedicated test functions for each test case here.

Reply:
indeed, it's kind hard to read.